### PR TITLE
Statements LOC info should not be modified when the debugger is enabled

### PR DESCRIPTION
**Pull request by rerobika**
The line info for each breakpoint can be calculated from the LOC index.
This patch resolves the problem, mentioned in #649.

Signed-off-by: Robert Fancsik <frobert@inf.u-szeged.hu>

### CODE REVIEW

File: src/interpreter/ByteCodeGenerator.cpp [hunk 1]
```diff
@@ -88,14 +88,36 @@ void ByteCodeGenerateContext::morphJumpPositionIntoComplexCase(ByteCodeBlock* cb
...
-void ByteCodeGenerateContext::insertBreakpoint(size_t line, Node* node)
+void ByteCodeGenerateContext::insertBreakpoint(size_t index, Node* node)
 {
     ASSERT(m_breakpointContext != nullptr);
 
-    if (line != 0 && line != m_breakpointContext->m_lastBreakpointLine) {
-        m_breakpointContext->m_breakpointLocations.push_back(Debugger::BreakpointLocation(line, (uint32_t)m_byteCodeBlock->currentCodeSize()));
+    if (index == SIZE_MAX) {
+        return;
+    }
```
Comment (clover2123) [inline src/interpreter/ByteCodeGenerator.cpp:97]: Is there any case that `index` has invalid value (SIZE_MAX)?

File: src/interpreter/ByteCodeGenerator.cpp [hunk 2]
```diff
@@ -88,14 +88,36 @@ void ByteCodeGenerateContext::morphJumpPositionIntoComplexCase(ByteCodeBlock* cb
...
+                    i++;
+                }
+                m_breakpointContext->m_lastBreakpointLine++;
+            }
+        }
+
+        m_breakpointContext->m_lastBreakpointIndex = index;
+    }
+
+    if (m_breakpointContext->m_lastBreakpointLine == sourceElementStart.line || lastLine != m_breakpointContext->m_lastBreakpointLine) {
```
Comment (clover2123) [inline src/interpreter/ByteCodeGenerator.cpp:118]: Why is `m_breakpointContext->m_lastBreakpointLine == sourceElementStart.line` added for this if statement?
IMO `lastLine != m_breakpointContext->m_lastBreakpointLine` check would be enough.

File: src/interpreter/ByteCodeGenerator.cpp [hunk 3]
```diff
@@ -88,14 +88,36 @@ void ByteCodeGenerateContext::morphJumpPositionIntoComplexCase(ByteCodeBlock* cb
...
+    StringView src = m_codeBlock->asInterpretedCodeBlock()->src();
+    ExtendedNodeLOC sourceElementStart = m_codeBlock->asInterpretedCodeBlock()->functionStart();
+    size_t lastLine = m_breakpointContext->m_lastBreakpointLine;
+
+    if (index > m_breakpointContext->m_lastBreakpointIndex) {
+        for (size_t i = m_breakpointContext->m_lastBreakpointIndex; i < index; i++) {
+            char16_t c = src.charAt(i - sourceElementStart.index);
+            if (EscargotLexer::isLineTerminator(c)) {
+                // skip \r\n
+                if (c == 13 && (i + 1 < index) && src.charAt(i + 1) == 10) {
```
Comment (clover2123) [inline src/interpreter/ByteCodeGenerator.cpp:108]: How about adding `UNLIKELY` macro for this if statement?

File: src/interpreter/ByteCodeGenerator.cpp [hunk 4]
```diff
@@ -151,7 +173,8 @@ ByteCodeBlock* ByteCodeGenerator::generateByteCode(Context* c, InterpretedCodeBl
 
 #ifdef ESCARGOT_DEBUGGER
     ByteCodeBreakpointContext breakpointContext;
-    breakpointContext.m_lastBreakpointLine = 0;
+    breakpointContext.m_lastBreakpointLine = codeBlock->asInterpretedCodeBlock()->functionStart().line;
+    breakpointContext.m_lastBreakpointIndex = codeBlock->asInterpretedCodeBlock()->functionStart().index;
```
Comment (clover2123) [inline src/interpreter/ByteCodeGenerator.cpp:177]: I'm wondering that initialization of each location info (`m_lastBreakpointLine`, `m_lastBreakpointIndex`) just to zero would have any issue. If this approach is available, then default constructor of `ByteCodeBreakpointContext` looks more proper at here.

File: src/interpreter/ByteCodeGenerator.cpp (hunk 2 above)
Comment (rerobika) [reply src/interpreter/ByteCodeGenerator.cpp]: I've slightly updated this case, bit there is one corner case:

```js
let j = function () {};

j();
```

Whenever you step in `j()` at least one breakpoint is needed in the function. However,  `lastLine` and `m_breakpointContext->m_lastBreakpointLine` would be 0 at the same time, but the breakpoint is needed, thats why the extra `lastLine == 0` check is needed.

**Comment by rerobika** [comment]
@clover2123 Thanks for the review, I updated the patch according to it.

File: src/interpreter/ByteCodeGenerator.cpp [hunk 5]
```diff
@@ -151,7 +171,8 @@ ByteCodeBlock* ByteCodeGenerator::generateByteCode(Context* c, InterpretedCodeBl
 
 #ifdef ESCARGOT_DEBUGGER
     ByteCodeBreakpointContext breakpointContext;
-    breakpointContext.m_lastBreakpointLine = 0;
+    // breakpointContext.m_lastBreakpointLine = codeBlock->asInterpretedCodeBlock()->functionStart().line;
+    // breakpointContext.m_lastBreakpointIndex = codeBlock->asInterpretedCodeBlock()->functionStart().index;
```
Comment (zherczeg) [inline src/interpreter/ByteCodeGenerator.cpp:175]: Why these lines are added as commented?

File: src/interpreter/ByteCodeGenerator.cpp [hunk 6]
```diff
@@ -88,14 +88,34 @@ void ByteCodeGenerateContext::morphJumpPositionIntoComplexCase(ByteCodeBlock* cb
...
+    ExtendedNodeLOC sourceElementStart = m_codeBlock->asInterpretedCodeBlock()->functionStart();
+    size_t lastLine = m_breakpointContext->m_lastBreakpointLine;
+    index -= sourceElementStart.index;
+
+    if (index > m_breakpointContext->m_lastBreakpointIndex) {
+        for (size_t i = m_breakpointContext->m_lastBreakpointIndex; i < index; i++) {
+            char16_t c = src.charAt(i);
+            if (EscargotLexer::isLineTerminator(c)) {
+                // skip \r\n
+                if (UNLIKELY (c == 13 && (i + 1 < index) && src.charAt(i + 1) == 10)) {
```
Comment (zherczeg) [inline src/interpreter/ByteCodeGenerator.cpp:106]: Don't we have any constants for newline?

File: src/parser/esprima_cpp/esprima.cpp [hunk 7]
```diff
@@ -4521,8 +4485,9 @@ class Parser {
         bool isEndedWithReturnNode = referNode && referNode->type() == ASTNodeType::ReturnStatement;
         if (!isEndedWithReturnNode) {
             referNode = body->appendChild(builder.createReturnStatementNode(nullptr));
+
 #ifdef ESCARGOT_DEBUGGER
-            referNode->setBreakpointInfo(this->lookahead.start, this->lookahead.lineNumber);
+            referNode->m_loc.index = this->createNode().index;
```
Comment (zherczeg) [inline src/parser/esprima_cpp/esprima.cpp:4490]: Do we have to create a node to access the index?

File: src/interpreter/ByteCodeGenerator.cpp (hunk 6 above)
Comment (rerobika) [reply src/interpreter/ByteCodeGenerator.cpp:106]: No we don't.

File: src/parser/esprima_cpp/esprima.cpp [hunk 8]
```diff
@@ -4001,16 +3970,18 @@ class Parser {
             this->throwError(Messages::IllegalReturn);
         }
 
+        MetaNode node = this->createNode();
         this->expectKeyword(ReturnKeyword);
 
         bool hasArgument = !this->match(SemiColon) && !this->match(RightBrace) && !this->hasLineTerminator && this->lookahead.type != EOFToken;
         ASTNode argument = nullptr;
         if (hasArgument) {
+            node = this->createNode();
```
Comment (clover2123) [inline src/parser/esprima_cpp/esprima.cpp:3979]: Why this line is added?
IMO the firstly created node seems enough for location of return statement.

File: src/interpreter/ByteCodeGenerator.cpp (hunk 2 above)
Comment (clover2123) [reply src/interpreter/ByteCodeGenerator.cpp]: I'm little bit confused in line start number (correct me if I'm wrong)
AFAIK line location starts with 1 not 0, so the extra check seems unnecessary.

File: src/parser/esprima_cpp/esprima.cpp [hunk 9]
```diff
@@ -5854,14 +5819,11 @@ class Parser {
 
         MetaNode endNode = this->createNode();
         this->currentScopeContext->m_bodyEndLOC.index = endNode.index;
-#if !(defined NDEBUG) || defined ESCARGOT_DEBUGGER
+#if !(defined NDEBUG)
         this->currentScopeContext->m_bodyEndLOC.line = endNode.line;
         this->currentScopeContext->m_bodyEndLOC.column = endNode.column;
 #endif
```
Comment (clover2123) [inline src/parser/esprima_cpp/esprima.cpp:5825]: If debugger does not use the ExtendedNodeLOC for `m_bodyEndLOC`, then other part related to it should be fixed too.
The below is an example.
https://github.com/Samsung/escargot/blob/b977ab7e54a10fdeac7d618e0a90308f7d723709/src/parser/ast/ASTContext.h#L215-L217

File: src/interpreter/ByteCodeGenerator.cpp (hunk 2 above)
Comment (rerobika) [reply src/interpreter/ByteCodeGenerator.cpp]: That check is still not unnecessary, only the appellation of the variable is a bit misleading. I'm going to rename it, since now it stores the **offset** from the last line/index not the last line/index.

**Comment by rerobika** [comment]
@clover2123 Thanks for the review, the PR is updated.

File: src/interpreter/ByteCodeGenerator.cpp (hunk 2 above)
Comment (clover2123) [reply src/interpreter/ByteCodeGenerator.cpp]: I'm still wondering why `lineOffset` is used to calculate the `lineNumber` instead of directly using the `lineNumber`.
For the following corner case,
```js
let j = function () {};
j();
```
When we step inside `j()`, one breakpoint is triggered because every empty function has one hidden return statement in itself. And the breakpoint of the return statement has of course the same line number of `let j` but this breakpoint is inserted during the bytecode generation of `function j `not global code generation. Since `ByteCodeBreakpointContext` is initialized at each bytecode generation process, I think that `lastLine != m_breakpointContext->m_lastBreakpointLine` check would be enough.

Comment (rerobika) [reply src/interpreter/ByteCodeGenerator.cpp]: Let me revise myself a little bit after some investigation. Whenever the bytecodeGeneration of `function j(){}` start the `m_lastBreakpointLineOffset` is zeroed out in the initialization on the new BytecodeGeneratorContext. Yes there is a manually inserted return node, with a specified node location **index**. Then the lookup starts to find the correspoding line for this specified index and because there is no `\r\n` in the empty function body the line number is not increased. So `lastLineOffset` remains zero and the context's lastLineOffset is also remains zero since it is not increased by the lookup. If we don't check for `lastLineOffset != 0` the `lastLineOffset != m_breakpointContext->m_lastBreakpointLineOffset` compares 0 with 0 which results no breakpoint insertion.

Comment (clover2123) [reply src/interpreter/ByteCodeGenerator.cpp]: Thank you for your neat explanation!
Now I have some minor suggestions (you can ignore it if you don't mind)
- How about storing line/index number instead of offset?

File: src/interpreter/ByteCodeGenerator.cpp [hunk 10]
```diff
@@ -88,14 +88,34 @@ void ByteCodeGenerateContext::morphJumpPositionIntoComplexCase(ByteCodeBlock* cb
...
 {
     ASSERT(m_breakpointContext != nullptr);
+    ASSERT(index != SIZE_MAX);
+
+    StringView src = m_codeBlock->asInterpretedCodeBlock()->src();
+    ExtendedNodeLOC sourceElementStart = m_codeBlock->asInterpretedCodeBlock()->functionStart();
+    size_t lastLineOffset = m_breakpointContext->m_lastBreakpointLineOffset;
+    index -= sourceElementStart.index;
+
+    if (index > m_breakpointContext->m_lastBreakpointIndexOffset) {
```
Comment (clover2123) [inline src/interpreter/ByteCodeGenerator.cpp:101]: IMO index would be increased sequentially, so this check code seems always true.

File: src/interpreter/ByteCodeGenerator.cpp [hunk 11]
```diff
@@ -88,14 +88,34 @@ void ByteCodeGenerateContext::morphJumpPositionIntoComplexCase(ByteCodeBlock* cb
...
+        for (size_t i = m_breakpointContext->m_lastBreakpointIndexOffset; i < index; i++) {
+            char16_t c = src.charAt(i);
+            if (EscargotLexer::isLineTerminator(c)) {
+                // skip \r\n
+                if (UNLIKELY(c == 13 && (i + 1 < index) && src.charAt(i + 1) == 10)) {
+                    i++;
+                }
+                m_breakpointContext->m_lastBreakpointLineOffset++;
```
Comment (clover2123) [inline src/interpreter/ByteCodeGenerator.cpp:109]: Instead of directly increasing `m_lastBreakpointLineOffset`, how about using a temporal variable and saving it's result into `m_lastBreakpointLineOffset`?

File: src/interpreter/ByteCodeGenerator.cpp [hunk 12]
```diff
@@ -88,14 +88,34 @@ void ByteCodeGenerateContext::morphJumpPositionIntoComplexCase(ByteCodeBlock* cb
...
+            }
+        }
+
+        m_breakpointContext->m_lastBreakpointIndexOffset = index;
+    }
 
-    if (line != 0 && line != m_breakpointContext->m_lastBreakpointLine) {
-        m_breakpointContext->m_breakpointLocations.push_back(Debugger::BreakpointLocation(line, (uint32_t)m_byteCodeBlock->currentCodeSize()));
+    if (lastLineOffset == 0 || lastLineOffset != m_breakpointContext->m_lastBreakpointLineOffset) {
```
Comment (clover2123) [inline src/interpreter/ByteCodeGenerator.cpp:116]: How about simplifying this check code like `lastLine > m_breakpointContext->m_lastBreakpointLine`?
For the first breakpoint insertion, we can also handle this corner case by setting the initial value.
For example,
```suggestion
    size_t lastLine = m_breakpointContext->m_lastBreakpointLine == 0 ? sourceElementStart.line : m_breakpointContext->m_lastBreakpointLine;
    ...
    if (lastLine > m_breakpointContext->m_lastBreakpointLine) {
```

File: src/interpreter/ByteCodeGenerator.cpp (hunk 2 above)
Comment (rerobika) [reply src/interpreter/ByteCodeGenerator.cpp]: I also had a version where I stored the line/index directly in the ByteCodeGenerator context, so I'd insist to the currently solution.

**Comment by rerobika** [comment]
@clover2123 Thanks for the review, I've updated the PR with a bit different approach to reduce the cost of extra checks.

File: src/parser/ast/ImplicitReturnStatementNode.h [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    {
+    }
+
+    virtual ASTNodeType type() override { return ASTNodeType::ImplicitReturnStatement; }
+    virtual void generateStatementByteCode(ByteCodeBlock* codeBlock, ByteCodeGenerateContext* context) override
+    {
+        ASSERT(context->tryCatchWithBlockStatementCount() == 0);
+#ifdef ESCARGOT_DEBUGGER
+        ASSERT(context->m_breakpointContext->m_lastBreakpointLineOffset == 0);
+        context->insertBreakpointAt(context->m_codeBlock->asInterpretedCodeBlock()->functionStart().line, this);
```
Comment (clover2123) [inline src/parser/ast/ImplicitReturnStatementNode.h:41]: How about merging this part into the original `ReturnStatementNode` instead of adding new `ImplicitReturnStatementNode`?
It seems that we can recognize each implicit return statement by checking `context->m_breakpointContext->m_lastBreakpointLineOffset == 0` or `codeBlock->currentCodeSize() == 0`.

File: src/parser/ast/ProgramNode.h [hunk 14]
```diff
@@ -54,7 +54,11 @@ class ProgramNode : public StatementNode {
 
 #ifdef ESCARGOT_DEBUGGER
         if (context->m_breakpointContext->m_breakpointLocations.size() == 0) {
-            insertBreakpoint(context);
+            if (codeBlock->m_isEvalMode) {
+                context->insertBreakpointAt(1, this);
```
Comment (clover2123) [inline src/parser/ast/ProgramNode.h:58]: It's just a question. Why is this part added especially for eval code?

File: src/parser/ast/ImplicitReturnStatementNode.h (hunk 13 above)
Comment (rerobika) [reply src/parser/ast/ImplicitReturnStatementNode.h:41]: Good idea! That's what I've been looking for.

File: src/parser/ast/ProgramNode.h (hunk 14 above)
Comment (rerobika) [reply src/parser/ast/ProgramNode.h:58]: Due to pending breakpoints you are able to set a breakpoint into a function created in eval like this:

```js

-> var a = 5;
   eval ("function f(){}");
```
debugger:
```
break f
pending? yes
```

Also whenever the eval code is processed no exact line info exist, also there breakpoint emitting does not happens since there is only a single function, but each parsed script needs at least 1 breakpoint. That's why the manually inserted breakpoint is needed.

**Comment by rerobika** [comment]
@clover2123 Please check it again, I think we can finally deal with the current solution.

File: src/parser/esprima_cpp/esprima.cpp [hunk 15]
```diff
@@ -4523,10 +4486,19 @@ class Parser {
...
         if (!isEndedWithReturnNode) {
-            referNode = body->appendChild(builder.createReturnStatementNode(nullptr));
+            Node* returnNode;
 #ifdef ESCARGOT_DEBUGGER
-            referNode->setBreakpointInfo(this->lookahead.start, this->lookahead.lineNumber);
+            if (!referNode) {
+                returnNode = builder.createReturnStatementNode(nullptr);
+            } else {
+                returnNode = builder.createReturnStatementNode(nullptr);
+            }
```
Comment (zherczeg) [inline src/parser/esprima_cpp/esprima.cpp:4495]: What is the difference in these two cases?

Comment (rerobika) [reply src/parser/esprima_cpp/esprima.cpp]: Thanks, I've forgotten to remove it.
